File: src/main/java/com/huythanh0x/training_thanhvh_java_spring_jwt_jpa/repository/CouponCourseRepository.java
package com.huythanh0x.training_thanhvh_java_spring_jwt_jpa.repository;

import com.huythanh0x.training_thanhvh_java_spring_jwt_jpa.model.coupon.CouponCourseData;
import jakarta.transaction.Transactional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.Set;

@Repository
public interface CouponCourseRepository extends JpaRepository<CouponCourseData, Integer> {

    Page<CouponCourseData> findByTitleContainingOrDescriptionContainingOrHeadingContaining(String title, String description, String heading, Pageable pageable);

    Page<CouponCourseData> findByRatingGreaterThanAndContentLengthGreaterThanAndLevelContainingAndCategoryIsContainingIgnoreCaseAndLanguageContaining(float rating, int contentLength, String level, String category, String language, Pageable pageable);

    CouponCourseData findByCourseId(Integer courseId);

    @Modifying
    @Transactional
    @Query("DELETE FROM CouponCourseData ccd WHERE ccd.couponUrl IN :expiredCouponUrls")
    void deleteAllCouponsByUrl(Set<String> expiredCouponUrls);

    @Modifying
    @Transactional
    void deleteByCouponUrl(String couponUrl);
}

File: modules/coupon-domain/src/main/kotlin/com/thanh0x/coursedeal/repository/CouponCourseRepository.kt
package com.thanh0x.coursedeal.repository

import com.thanh0x.coursedeal.model.coupon.CouponCourseData
import jakarta.transaction.Transactional
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.time.Instant
import java.time.LocalDateTime

@Repository
interface CouponCourseRepository : JpaRepository<CouponCourseData, Int> {
    fun findByTitleContainingOrDescriptionContainingOrHeadingContaining(
        title: String,
        description: String,
        heading: String,
        pageable: Pageable,
    ): Page<CouponCourseData>

    @Query(
        """
        SELECT c FROM CouponCourseData c WHERE 
        c.rating > :rating AND 
        c.contentLength > :contentLength AND 
        c.level LIKE %:level% AND 
        LOWER(c.category) LIKE LOWER(CONCAT('%', :category, '%')) AND 
        c.language LIKE %:language%
        """,
    )
    @Suppress("LongParameterList")
    fun findWithStructuredFilters(
        @Param("rating") rating: Float,
        @Param("contentLength") contentLength: Int,
        @Param("level") level: String,
        @Param("category") category: String,
        @Param("language") language: String,
        pageable: Pageable,
    ): Page<CouponCourseData>

    fun findByCourseId(courseId: Int): CouponCourseData?

    @Modifying
    @Transactional
    @Query("DELETE FROM CouponCourseData ccd WHERE ccd.couponUrl IN :expiredCouponUrls")
    fun deleteAllCouponsByUrl(
        @Param("expiredCouponUrls") expiredCouponUrls: Set<String>,
    )

    @Modifying
    @Transactional
    fun deleteByCouponUrl(couponUrl: String)

    @Query("SELECT c.couponUrl FROM CouponCourseData c WHERE c.couponUrl IS NOT NULL")
    fun findAllCouponUrls(): Set<String>

    fun findByCouponUrl(couponUrl: String): CouponCourseData?

    @Query("SELECT c.courseId FROM CouponCourseData c WHERE c.couponUrl = :couponUrl")
    fun findCourseIdByCouponUrl(
        @Param("couponUrl") couponUrl: String,
    ): Int?

    @Query(
        """
        SELECT DISTINCT c.couponUrl FROM CouponCourseData c WHERE 
        c.couponUrl IS NOT NULL AND (
        (c.expiredDate <= :expirationThreshold) OR 
        (c.usesRemaining < :minUsesRemaining) OR 
        (c.updatedAt < :updatedBefore)
        )
        """,
    )
    fun findCouponUrlsNeedingRefresh(
        @Param("expirationThreshold") expirationThreshold: Instant,
        @Param("minUsesRemaining") minUsesRemaining: Int,
        @Param("updatedBefore") updatedBefore: LocalDateTime,
    ): Set<String>
}
